File: ios/GradeThread/Passport/PassportService.swift
import Foundation

/// Resolves an inventory item to its Garment Passport and reads the public,
/// PII-free provenance chain (US-1115). The passport identity chain is:
///
///   inventory_items.grade_report_id → grade_reports.garment_id
///                                    → garments.public_passport_slug
///
/// The two intermediate reads go through Supabase (owner-readable under RLS,
/// same as ``ItemGradeReportService``); the timeline + claim-handoff calls go
/// through the edge (the public read is PII-free + unauthenticated, the mint is
/// authed + workspace-scoped via ``EdgeAPI``).
@MainActor
enum PassportService {

    /// A resolved passport handle for an item: the public slug (for the share
    /// URL + timeline read) and the garment id (for the owner claim-handoff mint).
    struct Ref: Equatable {
        let slug: String
        let garmentId: String
    }

    // Supabase row shapes (snake_case — the Supabase client decoder does NOT
    // convert keys, matching ItemGradeReportService).
    private struct ItemLink: Decodable { let grade_report_id: String? }
    private struct ReportLink: Decodable { let garment_id: String? }
    private struct GarmentRow: Decodable { let id: String; let public_passport_slug: String }

    /// Resolve the passport ref for an inventory item, or nil when the item has
    /// no passport yet (ungraded, or a legacy grade not yet backfilled).
    static func resolve(inventoryItemId: String) async throws -> Ref? {
        let links: [ItemLink] = try await SupabaseShared.client
            .from("inventory_items")
            .select("grade_report_id")
            .eq("id", value: inventoryItemId)
            .limit(1)
            .execute()
            .value
        guard let reportId = links.first?.grade_report_id else { return nil }

        let reports: [ReportLink] = try await SupabaseShared.client
            .from("grade_reports")
            .select("garment_id")
            .eq("id", value: reportId)
            .limit(1)
            .execute()
            .value
        guard let garmentId = reports.first?.garment_id else { return nil }

        let garments: [GarmentRow] = try await SupabaseShared.client
            .from("garments")
            .select("id, public_passport_slug")
            .eq("id", value: garmentId)
            .limit(1)
            .execute()
            .value
        guard let g = garments.first else { return nil }
        return Ref(slug: g.public_passport_slug, garmentId: g.id)
    }

    /// Read the public, PII-free provenance chain for a slug from the edge.
    static func fetchTimeline(slug: String) async throws -> PassportTimeline {
        let encoded = slug.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? slug
        return try await EdgeAPI.shared.getJSON("/api/passport/\(encoded)")
    }

    /// Mint a single-use, time-bounded ownership-claim link the owner can hand
    /// off to a buyer (so they can claim the passport). Authed + workspace-scoped
    /// via ``EdgeAPI``; the raw link is returned exactly once.
    static func mintHandoffLink(garmentId: String) async throws -> PassportHandoff {
        struct Empty: Encodable {}
        let encoded = garmentId.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? garmentId
        let response: HandoffResponse = try await EdgeAPI.shared.postJSON(
            "/api/passport/garments/\(encoded)/claim-token",
            body: Empty()
        )
        guard let url = URL(string: response.claimUrl) else {
            throw EdgeAPIError.decoding("Malformed claim URL")
        }
        return PassportHandoff(url: url, expiresAt: response.expiresAt)
    }

    // EdgeAPI decoder is convertFromSnakeCase, so claim_url/expires_at arrive
    // camelCased.
    private struct HandoffResponse: Decodable {
        let token: String
        let claimUrl: String
        let expiresAt: String
    }
}

/// A minted ownership-claim handoff link plus its expiry (raw ISO string).
struct PassportHandoff: Equatable {
    let url: URL
    let expiresAt: String
}

File: android/app/src/main/java/com/gradethread/app/passport/PassportService.kt
package com.gradethread.app.passport

import com.gradethread.app.platform.net.EdgeApi
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.net.URLEncoder
import javax.inject.Inject
import javax.inject.Named
import javax.inject.Singleton

@Serializable
private data class ItemLink(@SerialName("grade_report_id") val gradeReportId: String? = null)

@Serializable
private data class ReportLink(@SerialName("garment_id") val garmentId: String? = null)

@Serializable
private data class GarmentRow(
    val id: String = "",
    @SerialName("public_passport_slug") val publicPassportSlug: String = "",
)

/**
 * The two identifiers the passport surface needs.
 *
 * [slug] reads the public chain; [garmentId] is what the owner-only claim mint
 * takes. The resolve walk already reads both off the same row, and dropping the
 * id meant the handoff had no id to work from at all.
 */
data class PassportRef(val garmentId: String, val slug: String)

/** A minted claim link. The raw token exists only here and is never stored. */
@Serializable
data class PassportHandoff(
    val token: String = "",
    @SerialName("claim_url") val claimUrl: String = "",
    @SerialName("expires_at") val expiresAt: String = "",
)

/**
 * US-1376: resolve an item to its passport, then read the public chain.
 *
 * The identity chain is three hops:
 *
 *   inventory_items.grade_report_id → grade_reports.garment_id
 *                                   → garments.public_passport_slug
 *
 * The intermediate reads go through the RLS-scoped client (owner-readable);
 * the timeline itself is the PUBLIC, PII-free edge endpoint. Nothing here needs
 * the seller's identity, and nothing here asks for it.
 */
interface PassportProviding {
    /** The passport ids for an item, or null when it has none yet. */
    suspend fun resolve(inventoryItemId: String): PassportRef?

    suspend fun timeline(slug: String): PassportTimeline

    /**
     * US-2494: mint a single-use ownership-claim link for the buyer.
     *
     * The server verifies the garment belongs to this workspace before minting
     * and stores only a HASH, so the raw link comes back exactly once and can
     * never be fetched again.
     */
    suspend fun mintClaimLink(garmentId: String): PassportHandoff
}

@Singleton
class PassportService @Inject constructor(
    private val client: SupabaseClient,
    @Named("shared") private val edge: EdgeApi,
) : PassportProviding {

    companion object {
        fun path(slug: String): String =
            "/api/passport/" + URLEncoder.encode(slug, "UTF-8").replace("+", "%20")

        fun claimTokenPath(garmentId: String): String =
            "/api/passport/garments/" +
                URLEncoder.encode(garmentId, "UTF-8").replace("+", "%20") +
                "/claim-token"

        private val json = Json { ignoreUnknownKeys = true }
    }

    /**
     * Null at any broken link in the chain, never an exception.
     *
     * An ungraded item, or a legacy grade that predates the passport backfill,
     * simply has no passport. That is an ordinary state for most of an
     * inventory, not a failure worth an error banner.
     */
    override suspend fun resolve(inventoryItemId: String): PassportRef? {
        val reportId = client.from("inventory_items")
            .select(Columns.raw("grade_report_id")) {
                filter { eq("id", inventoryItemId) }
                limit(1)
            }
            .decodeList<ItemLink>()
            .firstOrNull()
            ?.gradeReportId ?: return null

        val garmentId = client.from("grade_reports")
            .select(Columns.raw("garment_id")) {
                filter { eq("id", reportId) }
                limit(1)
            }
            .decodeList<ReportLink>()
            .firstOrNull()
            ?.garmentId ?: return null

        val row = client.from("garments")
            .select(Columns.raw("id, public_passport_slug")) {
                filter { eq("id", garmentId) }
                limit(1)
            }
            .decodeList<GarmentRow>()
            .firstOrNull() ?: return null

        if (row.id.isBlank() || row.publicPassportSlug.isBlank()) return null
        return PassportRef(garmentId = row.id, slug = row.publicPassportSlug)
    }

    override suspend fun timeline(slug: String): PassportTimeline = json.decodeFromString(
        PassportTimeline.serializer(),
        edge.getRaw(path(slug)),
    )

    override suspend fun mintClaimLink(garmentId: String): PassportHandoff = json.decodeFromString(
        PassportHandoff.serializer(),
        // No body: the garment is named in the path and the owner is the
        // caller's token. Nothing about the mint is decided by the client.
        edge.postRaw(claimTokenPath(garmentId), "{}"),
    )
}
